### ISSUE TITLE
Unable to build Realm Xcode 11 beta 5 using Swift PM for iOS app with WatchKit Extension

### ISSUE THREAD

**Issue by bmalbuck** [realm/realm-cocoa#6226]
## Steps to Reproduce

1. Using Xcode 11 beta 5, create a new "iOS App with Watch App" project under the watchOS tab in choosing a template. Keep the default options when creating.
2. Add realm-cocoa Swift PM: File -> Swift Packages -> Add Package Dependency...
3. Paste https://github.com/realm/realm-cocoa.git in the repository URL. When prompted for revision. Select Exact 3.17.3.
4. Once Realm and RealmCore resolve and master branch checked out, Add Realm and RealmSwift to "Link Binary With Libraries" for BOTH the iOS AND WatchKit Extension targets.
5. Attempt to build the project.

![Screen Shot 2019-08-11 at 8 56 32 AM](https://user-images.githubusercontent.com/1012096/62836441-aaf58e80-bc17-11e9-8cdc-603ce58da26a.png)

**Comment by bmalbuck** [comment]
I created a test project available here:
https://github.com/bmalbuck/TestRealmWatchOS.git

**Comment by jsflax** [comment]
When using Swift PM, please point to the branch tg/xcode-11b1– please let me know if that solves the issue.

**Comment by bmalbuck** [comment]
Pointing to branch tg/xcode-11-b1 failed because that branch has no Package.swift manifest. Also, that branch has already been merged into master upstream from 3.17.3.

**Comment by jsflax** [comment]
I've followed your steps and am unable to reproduce this issue. When you are adding the packing from SwiftPM, only add it to the iOS app at first. Then, to use Realm in your watch app, click on the WatchKit Extension target and embed RealmSwift manually.

<img width="870" alt="Screenshot 2019-08-11 at 17 58 57" src="https://user-images.githubusercontent.com/4356928/62836988-b9689800-bc61-11e9-833b-4deb90b58ac1.png">

**Comment by bmalbuck** [comment]
That's what I have been doing but also added the Realm (core) library to extension. I started over, this time, only adding RealmSwift to extension but I'm getting the same error as shown in screenshot.

**Comment by jsflax** [comment]
Can you send a screen shot of your Frameworks & Libraries for each of your
targets?

Ar Domh 11 Lún 2019 ag 22:17, scríobh Brian Buck <notifications@github.com>:

**Comment by virtuosovidit** [comment]
![Screenshot 2019-08-12 at 4 09 22 AM](https://user-images.githubusercontent.com/20414466/62840417-037c6880-bcb8-11e9-832a-8576ffe0c7c0.png)
![Screenshot 2019-08-12 at 4 15 14 AM](https://user-images.githubusercontent.com/20414466/62840422-0c6d3a00-bcb8-11e9-8027-5c522ee75bfb.png)

This setup builds successfully when building from the watchOS target. However, fails with the following error when building from the iOS Target:
![Screenshot 2019-08-12 at 4 18 45 AM](https://user-images.githubusercontent.com/20414466/62840434-49393100-bcb8-11e9-85d4-0cf167bc32c1.png)

--
Important to note that the watchOS app is this particular target, not the standalone watchOS App target:
 
![Screenshot 2019-08-12 at 4 19 39 AM](https://user-images.githubusercontent.com/20414466/62840445-7a196600-bcb8-11e9-855e-8155c7dbced2.png)

**Comment by virtuosovidit** [comment]
Hi @jsflax 
It'll be great to know of any possible workaround for this issue, or any acknowledgement if this is being patched by Realm. It's 16th August, and we're probably less than a month away from iOS 13; If this is not on Realm's current timeline, please let us know, so we can at least plan for it accordingly.

**Comment by jsflax** [comment]
Hi @virtuosovidit ,

I'm still having trouble replicating this. I will speak to a colleague and see if they can replicate.

**Comment by virtuosovidit** [comment]
Can you confirm the Version Rules in Swift Package Manager on your end?
And, Perhaps share a dummy project, that helps in setting up Realm for watchOS using SPM?

**Comment by tgoyne** [comment]
I can reproduce this in beta 6. SPM is failing to add the RealmCore public header directory to the search path when building for watchOS, but it does include it when building for iOS.

**Comment by tgoyne** [comment]
Building via xcodebuild rather than Xcode (with `DEVELOPER_DIR=/Applications/Xcode-beta.app/Contents/Developer xcodebuild -project TestRealmWatchOS.xcodeproj -scheme 'TestRealmWatchOS WatchKit App'`) seems to work correctly. Does this also work for your real app?

I think this may just be an Xcode bug, so I'll try to put together a minimal repro case to report to apple.

**Comment by bswaidner** [comment]
Having a similar issue on Xcode beta 5/6. Currently, we have a target that depends on Realm and RealmSwift via SPM. In both DEBUG and RELEASE environments, we have no issues. However, when building the target in a BETA environment, we receive the following error: `error: no such module 'RealmSwift'`

Building via xcodebuild does not solve the issue.

**Comment by tgoyne** [comment]
I've filed a bug report with apple for this as it's broken even in a completely trivial case and I can't find a way to work around it.

**Comment by tgoyne** [comment]
I'm currently working on making our build system build and package a proper universal xcframework. The following steps should give you a working xcframework for ios+watchos (using the `tg/xcframework` branch):

```
export DEVELOPER_DIR=/Applications/Xcode-beta.app/Contents/Developer
sh build.sh ios-swift
sh build.sh watchos-swift
find build/DerivedData -name 'RealmObjc.framework' \
    | sed 's/.*/-framework &/' \
    | xargs xcodebuild -create-xcframework -output build/RealmObjc.xcframework
find build/DerivedData -name 'RealmSwift.framework' \
    | sed 's/.*/-framework &/' \
    | xargs xcodebuild -create-xcframework -output build/RealmSwift.xcframework
```

This'll give you `RealmObjc.xcframework` and `RealmSwift.xcframework` in the `build` directory, which appear to work correctly in @bmalbuck's test project.

**Comment by cnstoll** [comment]
I'm having a similar issue that I think you mentioned in another issue thread, where building the xcframework works on one computer with the latest version of Xcode...and fails on another one :( Have you had any luck troubleshooting that issue?

**Comment by tgoyne** [comment]
Which step is failing? If you're getting errors about not being able to find the destination when trying to build the xcframework, that's related to switching between Xcode versions. You can run `scripts/reset-simulators.rb` to delete all simulators and recreate them with the active version of Xcode if that's the problem.

**Comment by virtuosovidit** [comment]
@tgoyne I keep getting the following error when trying to build xcframework from the above steps: `error: unable to read the file at '/Users/.../realm-cocoa/build/DerivedData/Realm/Build/Products/Release-watchos/RealmObjc.framework/RealmObjc'`

**Comment by tgoyne** [comment]
https://github.com/realm/realm-cocoa/pull/6238 adds much more robust and tested support for building xcframeworks. `sh build.sh xcframework` gives you xcframeworks with support for all platforms, or you can do `sh build.sh xcframework ios watchos` for just those platforms. Note that you'll need to depend on `RealmObjc.xcframework` and `RealmSwift.xcframework` and not `Realm.xcframework`.

I have not been able to find a workaround for the swiftpm issue. The problem unfortunately seems to be in the closed-source Xcode part of it rather than the open-source spm libraries.

**Comment by virtuosovidit** [comment]
@tgoyne Just tried the approach mentioned in #6238 and I get the following errors with a build failed message:
`The following build commands failed:
	CompileC /Users/.../realm-cocoa/build/DerivedData/Realm/Build/Intermediates.noindex/Realm.build/Release/Realm.build/Objects-normal/x86_64/weak_realm_notifier.o /Users/.../realm-cocoa/Realm/ObjectStore/src/impl/weak_realm_notifier.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
	CompileC /Users/.../realm-cocoa/build/DerivedData/Realm/Build/Intermediates.noindex/Realm.build/Release/Realm.build/Objects-normal/x86_64/system_configuration.o /Users/.../realm-cocoa/Realm/ObjectStore/src/sync/impl/apple/system_configuration.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
	CompileC /Users/.../realm-cocoa/build/DerivedData/Realm/Build/Intermediates.noindex/Realm.build/Release/Realm.build/Objects-normal/x86_64/uuid.o /Users/.../realm-cocoa/Realm/ObjectStore/src/util/uuid.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
	CompileC /Users/.../realm-cocoa/build/DerivedData/Realm/Build/Intermediates.noindex/Realm.build/Release/Realm.build/Objects-normal/x86_64/transact_log_handler.o /Users/.../realm-cocoa/Realm/ObjectStore/src/impl/transact_log_handler.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
	CompileC /Users/.../realm-cocoa/build/DerivedData/Realm/Build/Intermediates.noindex/Realm.build/Release/Realm.build/Objects-normal/x86_64/sync_permission.o /Users/.../realm-cocoa/Realm/ObjectStore/src/sync/sync_permission.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
	CompileC /Users/.../realm-cocoa/build/DerivedData/Realm/Build/Intermediates.noindex/Realm.build/Release/Realm.build/Objects-normal/x86_64/thread_safe_reference.o /Users/.../realm-cocoa/Realm/ObjectStore/src/thread_safe_reference.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
	CompileC /Users/.../realm-cocoa/build/DerivedData/Realm/Build/Intermediates.noindex/Realm.build/Release/Realm.build/Objects-normal/x86_64/sync_session.o /Users/.../realm-cocoa/Realm/ObjectStore/src/sync/sync_session.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
	CompileC /Users/.../realm-cocoa/build/DerivedData/Realm/Build/Intermediates.noindex/Realm.build/Release/Realm.build/Objects-normal/x86_64/sync_metadata.o /Users/.../realm-cocoa/Realm/ObjectStore/src/sync/impl/sync_metadata.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
	CompileC /Users/.../realm-cocoa/build/DerivedData/Realm/Build/Intermediates.noindex/Realm.build/Release/Realm.build/Objects-normal/x86_64/sync_user.o /Users/.../realm-cocoa/Realm/ObjectStore/src/sync/sync_user.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
	CompileC /Users/.../realm-cocoa/build/DerivedData/Realm/Build/Intermediates.noindex/Realm.build/Release/Realm.build/Objects-normal/x86_64/work_queue.o /Users/.../realm-cocoa/Realm/ObjectStore/src/sync/impl/work_queue.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
	CompileC /Users/.../realm-cocoa/build/DerivedData/Realm/Build/Intermediates.noindex/Realm.build/Release/Realm.build/Objects-normal/x86_64/sync_manager.o /Users/.../realm-cocoa/Realm/ObjectStore/src/sync/sync_manager.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler`

**Comment by tgoyne** [comment]
That's just a list of build commands that failed and the actual error messages are elsewhere in the build log.

**Comment by virtuosovidit** [comment]
@tgoyne Thanks! I did locate errors, and it seems like a bunch of file not found errors.

```
error: Build input file cannot be found: '/Users/.../realm-cocoa/Realm/ObjectStore/src/collection_notifications.cpp' (in target 'Realm' from project 'Realm')

error: Build input file cannot be found: '/Users/.../realm-cocoa/Realm/ObjectStore/src/results.cpp' (in target 'Realm' from project 'Realm')

error: Build input file cannot be found: '/Users/.../realm-cocoa/Realm/ObjectStore/src/sync/impl/apple/network_reachability_observer.cpp' (in target 'Realm' from project 'Realm')

error: Build input file cannot be found: '/Users/.../realm-cocoa/Realm/ObjectStore/src/list.cpp' (in target 'Realm' from project 'Realm')

error: Build input file cannot be found: '/Users/.../realm-cocoa/Realm/ObjectStore/src/impl/realm_coordinator.cpp' (in target 'Realm' from project 'Realm')

error: Build input file cannot be found: '/Users/.../realm-cocoa/Realm/ObjectStore/src/impl/apple/external_commit_helper.cpp' (in target 'Realm' from project 'Realm')

error: Build input file cannot be found: '/Users/.../realm-cocoa/Realm/ObjectStore/src/impl/results_notifier.cpp' (in target 'Realm' from project 'Realm')

error: Build input file cannot be found: '/Users/.../realm-cocoa/Realm/ObjectStore/src/index_set.cpp' (in target 'Realm' from project 'Realm')

error: Build input file cannot be found: '/Users/.../realm-cocoa/Realm/ObjectStore/src/util/fifo.cpp' (in target 'Realm' from project 'Realm')

error: Build input file cannot be found: '/Users/.../realm-cocoa/Realm/ObjectStore/src/impl/collection_notifier.cpp' (in target 'Realm' from project 'Realm')

error: Build input file cannot be found: '/Users/.../realm-cocoa/Realm/ObjectStore/src/impl/list_notifier.cpp' (in target 'Realm' from project 'Realm')

error: Build input file cannot be found: '/Users/.../realm-cocoa/Realm/ObjectStore/src/impl/apple/keychain_helper.cpp' (in target 'Realm' from project 'Realm')
```

**Comment by bmalbuck** [comment]
@tgoyne thanks for working out a solution on this branch. Let's hope Apple can fix SPM.

 I was able to use `./build.sh xcframework ios watchos` to successfully compile the the xcframeworks. After build I linked `RealmObjc.xcframework` and `RealmSwift.xcframework` in both my projects iOS and watch extension targets. Project builds but now a runtime crash.

`Property cannot be marked @objc because its type cannot be represented in Objective-C`

I have not changed any source including a swift file that declares all my Realm objects. This file imports RealmSwift. For example:

public class MyRealmObject: Object {
...
}

public class AnotherObject: Object {
  @objc dynamic var myObj: MyRealmObject? // <-- runtime error
}

**Comment by tgoyne** [comment]
Joy, another shitty limitation of xcframeworks. It seems that having an `@objc` class which inherits from a class defined in a ABI-stable swift module is only supported on iOS 13 (and macOS 10.15 etc.). If I mark the class as `@available(iOS 13.0.0, *)` or set the deployment target to iOS 13 it works; if I target anything lower and mark the class as `@objc` it tells me "explicit '@objc' on subclass of 'Object' requires iOS 13.0.0". Even when targeting iOS 13 automatic class discovery doesn't seem to work, but explicitly listing the types in `objectTypes` does at least.

I'm sort of impressed by how aggressively anti-backwards-compatible all the new Xcode 11 features are. I guess Apple really doesn't want us supporting older versions.

**Comment by virtuosovidit** [comment]
@tgoyne So, I added the xcframeworks to watchOS and they seem to work fine on the simulator. However, when tested on device, I get the following error: 

dyld: dependent dylib 
```
'@rpath/RealmSwift.framework/RealmSwift' not found for '/private/var/containers/Bundle/Application/XXXXX-XXXXX-XXXXX-XXXXX/XXXXX.app/PlugIns/XXXXX Extension.appex/XXXXX Extension', 

tried but didn't find: '/private/var/containers/Bundle/Application/XXXXX-XXXXX-XXXXX-XXXXX/XXXXX.app/PlugIns/XXXXX Extension.appex/Frameworks/RealmSwift.framework/RealmSwift' 

'/private/var/containers/Bundle/Application/XXXXX-XXXXX-XXXXX-XXXXX/XXXXX.app/PlugIns/XXXXX Extension.appex/../../Frameworks/RealmSwift.framework/RealmSwift' 

'@rpath/RealmSwift.framework/RealmSwift' '/System/Library/Frameworks/RealmSwift.framework/RealmSwift' 

```

**Comment by tgoyne** [comment]
You'll need to change "Do Not Embed" to "Embed & Sign" for RealmSwift.xcframework in "Frameworks, Libraries, and Embedded Content" on the General tab for the target.

**Comment by virtuosovidit** [comment]
@tgoyne Yes, I'm using the Embed & Sign setting for the frameworks. It still doesn't work.

**Comment by cnstoll** [comment]
I got this (sort of) working by requiring iOS 13 (not great) and building the XCFramework. But then as soon as I run on an iOS 13 device I get this error:

2019-09-07 15:52:45.611164-0500 Grocery[55598:12530955] *** Terminating app due to uncaught exception 'RLMException', reason: 'Object type 'RealmGroceryStoreSortingOperation' is not managed by the Realm. If using a custom `objectClasses` / `objectTypes` array in your configuration, add `RealmGroceryStoreSortingOperation` to the list of `objectClasses` / `objectTypes`.'

I didn't make any other changes so I'm pretty sure something's wrong but I'm not sure what.

**Comment by tgoyne** [comment]
`objc_copyClassList()` is broken when building in library evolution mode, and that's how we automatically discover the `RealmSwift.Object` subclasses. I filed a bug (FB7201126) but it's probably far too late to hope for it to be fixed. You can work around this by explicitly listing all of the classes you're using in the `objectTypes` property on the Realm configuration.

**Comment by bmalbuck** [comment]
@cnstoll @tgoyne confirming that I also received the same runtime error after setting deployment target to iOS 13 and watchOS 6.0. Interestingly, I was going to try and add my custom Realm objects/types to a realm configuration instead of using default Realm config (ex: Realm()) but got sidetracked. I will try the workaround soon and report back.

**Comment by cnstoll** [comment]
I guess I should just try it, but can we just build regular .frameworks for iOS and watchOS and include them directly if we don't want to deal with this? Will those still support iOS 13 and Xcode 11?

**Comment by tgoyne** [comment]
The only upside of xcframeworks over the old frameworks is simpler installation (especially with catalyst, which makes things a bit more complicated). It's still possible to use normal frameworks, and given all the problems that have popped up with xcframeworks I guess it'll still be the recommended way.

**Comment by bmalbuck** [comment]
@tgoyne if using regular frameworks rather than xcframework does it solve not requiring minimum deployment target of iOS 13? I'd like to support 12 if possible.

**Comment by tgoyne** [comment]
Yes. The iOS 13 requirement is imposed by swift library evolution mode, which is required for xcframeworks, but not for regular frameworks. We'd like to eventually switch to building with it enabled (it eliminates the need to build the framework with the same version of Xcode as the app), but certainly won't until the downsides are sorted out.

**Comment by virtuosovidit** [comment]
@tgoyne Any updates on this? Realm on Apple Watch using SwiftPM is still throwing the same errors, as of Xcode 11.3

**Comment by tgoyne** [comment]
No, I haven't gotten any updates to any of the bugs I filed with apple.

**Comment by tgoyne** [comment]
The bug I filed about not being able to use SPM for watchkit extensions is marked as fixed in Xcode 11.4.

**Comment by leemaguire** [comment]
Closing as Realm on SwiftPM now works with Apple Watch.
